src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ConfiguredProjectImplicitActivationTracking.cs

```diff
@@ -55,45 +57,25 @@ protected override void Dispose(bool disposing)
             _targetBlock.Complete();
         }
 
-        private Task OnActiveConfigurationsChanged(IProjectVersionedValue<IConfigurationGroup<ProjectConfiguration>> e)
+        private async Task OnActiveConfigurationsChanged(IProjectVersionedValue<ValueTuple<IProjectCapabilitiesSnapshot, IConfigurationGroup<ProjectConfiguration>>> e)
         {
-            if (IsDisposing || IsDisposed)
-                return Task.CompletedTask;
+            using var capabilitiesContext = ProjectCapabilitiesContext.CreateIsolatedContext(_project, e.Value.Item1);
 
-            bool nowActive = e.Value.Contains(_project.ProjectConfiguration);
+            bool isActive = e.Value.Item2.Contains(_project.ProjectConfiguration);
 
-            // Are there any changes for my configuration?
-            if (nowActive == _isActive)
-                return Task.CompletedTask;
+            // If we're not active, there are no future services to activate
+            IReadOnlyCollection<IImplicitlyActiveService> futureServices = isActive ? ImplicitlyActiveServices.Select(s => s.Value).ToList()
+                                                                                    : Array.Empty<IImplicitlyActiveService>();
 
-            if (nowActive)
-            {
-                return OnImplicitlyActivated();
-            }
-            else if (_isActive)
-            {
-                return OnImplicitlyDeactivated();
-            }
+            // Deactive currently "active" services that no longer applicable
+            IEnumerable<IImplicitlyActiveService> servicesToDeactivate = _activeServices.Except(futureServices);
+            await Task.WhenAll(servicesToDeactivate.Select(c => c.DeactivateAsync()));
 
-            return Task.CompletedTask;
-        }
-
-        private Task OnImplicitlyActivated()
-        {
-            _isActive = true;
-
-            IEnumerable<Task> tasks = ImplicitlyActiveServices.Select(c => c.Value.ActivateAsync());
-
-            return Task.WhenAll(tasks);
-        }
-
-        private Task OnImplicitlyDeactivated()
-        {
-            _isActive = false;
-
-            IEnumerable<Task> tasks = ImplicitlyActiveServices.Select(c => c.Value.DeactivateAsync());
+            // Activate "non-active" services that are now applicable
+            IEnumerable<IImplicitlyActiveService> servicesToActivate = futureServices.Except(_activeServices);
+            await Task.WhenAll(servicesToActivate.Select(c => c.ActivateAsync()));
```

Comment:
Can the activations here be made concurrent with the deactivations above? If it's safe to do so, it may complete more quickly.

Reply:
In theory, yes, however, this will result in an overlap of Roslyn projects for language service integration for the same project, which is untested and might result in more OOMs - especially in multi-targeted projects. I tried to maintain as much behavior as I could at this moment, and will hold off on that.